### PR TITLE
Fixed #27098 -- Deprecated DatabaseIntrospection.get_indexes

### PR DESCRIPTION
**Pull request by claudep**
Should be rebased after:
 * #7131 is committed
 * ~~Index type is added to get_constraints() output (for test_gin_index). akki is working on it.~~

### CODE REVIEW

**Comment by claudep** [comment]
Introspection of Raster field indexes is not yet supported by `get_constraints()`…

**Comment by timgraham** [comment]
It would be nice to document what changes are needed for third-party database backends.

**Comment by claudep** [comment]
I found the reason why raster index is not in get_constraints() output. The index query in get_constraints is making an inner join with the pg_attribute table on pg_index.indkey. As raster indexes are using an expression (`ST_ConvexHull("column")`), indkey is 0 which excludes the index line from the results.
@akki Any opinion on how to handle this issue (not only on the SQL level, but also how to present it in the constraints dict)?

**Comment by claudep** [comment]
Note: `pg_get_indexdef` allows to retrieve the complete index creation command.

**Comment by akki** [comment]
Even the `get_indexes` method was querying this way but get_indexes used to take the last column into account (which coincidentally used to be the corresponding column when the tests). This is a bug that was present in `get_indexes` but is hitting us now.
I guess we'll have to use `pg_get_indexdef`.

**Comment by claudep** [comment]
Tests are passing now.
@akki, would be nice to have a look, as I'm not sure I choose the better design.

**Comment by timgraham** [comment]
I'm not sure about the deprecation/backwards-compatibility plan. The main situation that comes to mind is a third-party database backend that needs to be adapted for the changes. In that sense I think the fact that inspect no longer calls `get_indexes()` should be called out in the release notes. I don't have a situation in mind where user code calls `DatabaseIntrospection.get_indexes()` such that a deprecation of that presumably private method is needed, but I suppose it doesn't hurt.

File: django/core/management/commands/inspectdb.py [hunk 1]
```diff
@@ -71,13 +71,12 @@ def strip_prefix(s):
...
                     try:
-                        indexes = connection.introspection.get_indexes(cursor, table_name)
-                    except NotImplementedError:
-                        indexes = {}
-                    try:
                         constraints = connection.introspection.get_constraints(cursor, table_name)
                     except NotImplementedError:
                         constraints = {}
+                    primary_key_column = connection.introspection.get_primary_key_column(cursor, table_name)
+                    unique_columns = [c['columns'][0] for c in constraints.values()
```
Comment (timgraham) [inline django/core/management/commands/inspectdb.py]: Use hanging indent:
```
unique_columns = [
    c['columns'][0] for c in constraints.values()
    ...
]

File: django/db/backends/postgresql/introspection.py [hunk 2]
```diff
@@ -185,16 +191,21 @@ def get_constraints(self, cursor, table_name):
                 "foreign_key": tuple(used_cols.split(".", 1)) if kind == "f" else None,
                 "check": kind == "c",
                 "index": False,
+                "definition": None,
```
Comment (timgraham) [inline django/db/backends/postgresql/introspection.py]: Should it be added to docstring?

File: docs/releases/1.11.txt [hunk 3]
```diff
@@ -504,3 +504,5 @@ Miscellaneous
 * For the ``PyLibMCCache`` cache backend, passing ``pylibmc`` behavior settings
   as top-level attributes of ``OPTIONS`` is deprecated. Set them under a
   ``behaviors`` key within ``OPTIONS`` instead.
+* ``DatabaseIntrospection.get_indexes()`` is deprecated.
```
Comment (timgraham) [inline docs/releases/1.11.txt]: deprecated in favor of...

File: django/db/backends/mysql/introspection.py [hunk 4]
```diff
@@ -122,6 +124,10 @@ def get_key_columns(self, cursor, table_name):
         return key_columns
 
     def get_indexes(self, cursor, table_name):
+        warnings.warn(
+            "get_indexes() is deprecated. Use get_constraints() instead.",
+            RemovedInDjango21Warning
```
Comment (timgraham) [inline django/db/backends/mysql/introspection.py]: stacklevel=2

File: django/db/backends/postgresql/introspection.py [hunk 5]
```diff
@@ -191,16 +191,21 @@ def get_constraints(self, cursor, table_name):
...
         cursor.execute("""
             SELECT
                 indexname, array_agg(attname), indisunique, indisprimary,
-                array_agg(ordering), amname
+                array_agg(ordering), amname, exprdef
             FROM (
                 SELECT
                     c2.relname as indexname, idx.*, attr.attname, am.amname,
                     CASE
+                        WHEN idx.indexprs IS NOT NULL THEN
```
Comment (akki) [inline django/db/backends/postgresql/introspection.py]: Sorry, I missed this PR.
From what I found [in the docs](https://www.postgresql.org/docs/9.5/static/catalog-pg-index.html), this looks like the right thing to do. Thanks :+1:

File: django/db/backends/base/introspection.py [hunk 6]
```diff
@@ -143,13 +141,14 @@ def get_primary_key_column(self, cursor, table_name):
...
-            if column[1]['primary_key']:
-                return column[0]
+        for constraint in self.get_constraints(cursor, table_name).values():
+            if constraint['primary_key']:
+                return constraint['columns'][0]
         return None
 
     def get_indexes(self, cursor, table_name):
         """
+        Deprecated in Django 1.11, use get_constraints instead.
```
Comment (claudep) [inline django/db/backends/base/introspection.py]: I don't see the need for that, as the removal of `get_indexes()` is already listed in the deprecation timeline (which should be our "check list" when removing deprecated stuff).

Comment (timgraham) [inline django/db/backends/base/introspection.py]: It's just a foolproof technique I've used in the past. Without the ability to grep for the deprecation class, there's more thinking involved when doing the removals.

Comment (claudep) [inline django/db/backends/base/introspection.py]: I can understand that in some cases, but here grepping for `def get_indexes` should be straightforward enough, IMHO.

File: docs/releases/1.11.txt [hunk 7]
```diff
@@ -533,3 +533,8 @@ Miscellaneous
   :ttag:`{% include %} <include>` template tag is deprecated as the behavior is
   often more confusing than helpful. In Django 2.1, the exception will be
   raised.
+
+* ``DatabaseIntrospection.get_indexes()`` is deprecated in favor of
+  ``DatabaseIntrospection.get_constraints()``. The :djadmin:`inspectdb` command
```
Comment (timgraham) [inline docs/releases/1.11.txt]: Please move the second sentence to "Database backend API" under backwards-incompatible changes, e.g. :djadmin:`inspectdb` no longer calls DatabaseIntrospection.get_indexes(). Add unique indexes to the return value of get_constraints().

File: django/db/backends/mysql/introspection.py [hunk 8]
```diff
@@ -122,6 +124,10 @@ def get_key_columns(self, cursor, table_name):
         return key_columns
 
     def get_indexes(self, cursor, table_name):
+        warnings.warn(
+            "get_indexes() is deprecated. Use get_constraints() instead.",
```
Comment (timgraham) [inline django/db/backends/mysql/introspection.py]: Doesn't matter much, but the usual wording is "in favor of".

File: docs/releases/1.11.txt [hunk 9]
```diff
@@ -392,6 +392,10 @@ Database backend API
   dictionaries with a list of ``'ASC'`` and/or ``'DESC'`` values corresponding
   to the the ordering of each column in the index.
 
+* :djadmin:`inspectdb` no longer calls ``DatabaseIntrospection.get_indexes()``
+  which is deprecated. Custom database backends should ensure all types of
+  indexes are returned by ``DatabaseIntrospection.get_constraints()``.
```
Comment (claudep) [inline docs/releases/1.11.txt]: @timgraham May I ask for a final review of this note?

Comment (timgraham) [inline docs/releases/1.11.txt]: Looks good.

**Pull request closed without merging**